fabric-input-events-v1/src/main/java/net/fabricmc/fabric/api/client/event/input/ClientInputEvents.java

```diff
@@ -0,0 +1,109 @@
+import net.fabricmc.fabric.api.event.EventFactory;
+
+@Environment(EnvType.CLIENT)
+public final class ClientInputEvents {
+	public static final Event<KeyState> KEY_PRESSED
+		= EventFactory.createSimpleArrayBacked(KeyEvent.class, KeyState.class,
+		listener -> key -> listener.accept(key),
+		listener -> key -> listener.onKey(key));
+	public static final Event<KeyState> KEY_RELEASED
+		= EventFactory.createSimpleArrayBacked(KeyEvent.class, KeyState.class,
+		listener -> key -> listener.accept(key),
+		listener -> key -> listener.onKey(key));
+	public static final Event<KeyState> KEY_REPEATED
+		= EventFactory.createSimpleArrayBacked(KeyEvent.class, KeyState.class,
+		listener -> key -> listener.accept(key),
+		listener -> key -> listener.onKey(key));
+	public static final Event<KeybindState> KEYBIND_PRESSED
+		= EventFactory.createSimpleArrayBacked(KeybindEvent.class, KeybindState.class,
+		listener -> key -> listener.accept(key),
+		listener -> key -> listener.onKeybind(key));
+	public static final Event<KeybindState> KEYBIND_RELEASED
+		= EventFactory.createSimpleArrayBacked(KeybindEvent.class, KeybindState.class,
+		listener -> key -> listener.accept(key),
+		listener -> key -> listener.onKeybind(key));
+	public static final Event<KeybindState> KEYBIND_REPEATED
+		= EventFactory.createSimpleArrayBacked(KeybindEvent.class, KeybindState.class,
+		listener -> key -> listener.accept(key),
+		listener -> key -> listener.onKeybind(key));
+	public static final Event<CharState> CHAR_TYPED
+		= EventFactory.createSimpleArrayBacked(CharEvent.class, CharState.class,
+		listener -> key -> listener.accept(key),
+		listener -> key -> listener.onChar(key));
+	public static final Event<MouseMove> MOUSE_MOVED
+		= EventFactory.createSimpleArrayBacked(MouseMoveEvent.class, MouseMove.class,
+		listener -> mouse -> listener.accept(mouse),
+		listener -> mouse -> listener.onMouseMoved(mouse));
+	public static final Event<MouseButtonState> MOUSE_BUTTON_PRESSED
+		= EventFactory.createSimpleArrayBacked(MouseButtonEvent.class, MouseButtonState.class,
+		listener -> mouse -> listener.accept(mouse),
+		listener -> mouse -> listener.onMouseButton(mouse));
+	public static final Event<MouseButtonState> MOUSE_BUTTON_RELEASED
+		= EventFactory.createSimpleArrayBacked(MouseButtonEvent.class, MouseButtonState.class,
+		listener -> mouse -> listener.accept(mouse),
+		listener -> mouse -> listener.onMouseButton(mouse));
+	public static final Event<MouseScroll> MOUSE_WHEEL_SCROLLED
+		= EventFactory.createSimpleArrayBacked(MouseScrollEvent.class, MouseScroll.class,
+		listener -> mouse -> listener.accept(mouse),
+		listener -> mouse -> listener.onMouseScrolled(mouse));
+	public static final Event<FileDrop> FILE_DROPPED
+		= EventFactory.createSimpleArrayBacked(String[].class, FileDrop.class,
+		listener -> paths -> listener.accept(paths),
+		listener -> paths -> listener.onFilesDropped(paths));
+
+	@FunctionalInterface
+	public interface KeyState {
+		void onKey(KeyEvent key);
```

Comment:
Okay so the big changes start here. Typically we would just pass the parameters here, such as key code, modifiers, etc

This avoids the need to needlessly allocate an object here.

Reply:
I might start an issue about this because I think this is the wrong way to go about things.

In this case, specifically, it is very useful to obtain additional information not originally provided by the event, such as getting the mouse position or buttons being held during a scroll event or `KeyEvent::getKey` and `KeyEvent::isPressed`.

But, I think it is good to limit the listeners to one input even in the general case because then events would map much more cleanly to monads/the observer pattern (without wrapper overhead).

Adhering to the observer pattern would also make the design of the future event generation codegen much easier if any would even be needed at all (Observers compose really well).

> This avoids the need to needlessly allocate an object here.

You can cache objects and use getter interfaces to prevent mutability. One function call of extra overhead but it has no allocation cost and makes the entire events API much more extensible (can add new inputs at any time). I don't think the overhead is even an issue for 99% of use cases (Can just call the function once and store the result in a local variable, everyone does that for `Minecraft.getInstance` already anways), and in the rare case we do need to call thousands of listeners per second we still have the option of providing "unsafe" events that provide the mutable event object as input (Under the contract that no listeners modify the input, of course). It's probably even possible to create a class with final members and then use ASM to modify the cached event object.

All that together would actually be faster because it doesn't have the same hidden cost of allocation stack frames and copying parameters over and over again.

Reply:
Actually, on the topic of mutability, aren't we already implicitly requiring listeners to not modify their inputs anyway?